internal/admin/admin_test.go

```diff
@@ -223,6 +226,295 @@ func TestAdmin_GetStorageNode(t *testing.T) {
+// is not sealed in the event of intermittent failure of the mr request in that situation.
+func TestAdmin_Storagenode_HeartbeatTimeout_Seal(t *testing.T) {
+	const (
+		cid         = types.ClusterID(1)
+		tpid        = types.TopicID(1)
+		victimsnid  = types.StorageNodeID(1)
+		normalsnid  = types.StorageNodeID(2)
+		victimlsid1 = types.LogStreamID(1)
+		victimlsid2 = types.LogStreamID(2)
+		normallsid1 = types.LogStreamID(3)
+		normallsid2 = types.LogStreamID(4)
+		//TODO:: it should be 2
+		expectedSealed = 1
+	)
+
+	victimsnpath := filepath.Join("/tmp", volume.StorageNodeDirName(cid, victimsnid))
+	normalsnpath := filepath.Join("/tmp", volume.StorageNodeDirName(cid, normalsnid))
+	createTime := time.Now().UTC()
+
+	var mrfail atomic.Bool
+	var failed atomic.Int32
+
+	victimlsids := set.New(2)
+	victimlsids.Add(victimlsid1)
+	victimlsids.Add(victimlsid2)
+
+	normallsids := set.New(2)
+	normallsids.Add(normallsid1)
+	normallsids.Add(normallsid2)
+
+	victimStorageNodeMetadata := &snpb.StorageNodeMetadataDescriptor{
+		ClusterID: cid,
+		StorageNode: varlogpb.StorageNode{
+			StorageNodeID: victimsnid,
+		},
+		LogStreamReplicas: []snpb.LogStreamReplicaMetadataDescriptor{
+			{
+				LogStreamReplica: varlogpb.LogStreamReplica{
+					StorageNode: varlogpb.StorageNode{
+						StorageNodeID: victimsnid,
+					},
+					TopicLogStream: varlogpb.TopicLogStream{
+						TopicID:     tpid,
+						LogStreamID: victimlsid1,
+					},
+				},
+			},
+			{
+				LogStreamReplica: varlogpb.LogStreamReplica{
+					StorageNode: varlogpb.StorageNode{
+						StorageNodeID: victimsnid,
+					},
+					TopicLogStream: varlogpb.TopicLogStream{
+						TopicID:     tpid,
+						LogStreamID: victimlsid2,
+					},
+				},
+			},
+		},
+		StartTime: createTime,
+	}
+
+	normalStorageNodeMetadata := &snpb.StorageNodeMetadataDescriptor{
+		ClusterID: cid,
+		StorageNode: varlogpb.StorageNode{
+			StorageNodeID: normalsnid,
+		},
+		LogStreamReplicas: []snpb.LogStreamReplicaMetadataDescriptor{
+			{
+				LogStreamReplica: varlogpb.LogStreamReplica{
+					StorageNode: varlogpb.StorageNode{
+						StorageNodeID: normalsnid,
+					},
+					TopicLogStream: varlogpb.TopicLogStream{
+						TopicID:     tpid,
+						LogStreamID: normallsid1,
+					},
+				},
+			},
+			{
+				LogStreamReplica: varlogpb.LogStreamReplica{
+					StorageNode: varlogpb.StorageNode{
+						StorageNodeID: normalsnid,
+					},
+					TopicLogStream: varlogpb.TopicLogStream{
+						TopicID:     tpid,
+						LogStreamID: normallsid2,
+					},
+				},
+			},
+		},
+		StartTime: createTime,
+	}
+
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	mock := newTestMock(ctrl)
+	mock.MockClusterMetadataView.EXPECT().ClusterMetadata(gomock.Any()).DoAndReturn(
+		func(ctx context.Context) (*varlogpb.MetadataDescriptor, error) {
+			if ctx.Err() != nil {
+				return nil, ctx.Err()
+			}
+
+			if mrfail.Load() && rand.Int32N(100) < 20 {
+				return nil, context.Canceled
+			}
```

Comment:
Can we change this more deterministically - for example, how about it removing probable cancellation?

Reply:
ClusterMetadata should fail during health checks, but not fail during seal processing. This is not easy to reproduce in tests using mock. So I created a test based on probability.

Reply:
Okay, I understood. 😄